superset/models/sql_lab.py

```diff
@@ -0,0 +1,159 @@
+"""A collection of ORM sqlalchemy models for SQL Lab"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import re
+from datetime import datetime
+
+from future.standard_library import install_aliases
+
+from flask import Markup
+
+from flask_appbuilder import Model
+
+import sqlalchemy as sqla
+from sqlalchemy import (
+    Column, Integer, String, ForeignKey, Text, Boolean,
+    DateTime, Numeric,
+)
+from sqlalchemy.orm import backref, relationship
+
+from superset.utils import QueryStatus
+from superset.models.helpers import AuditMixinNullable
+
+install_aliases()
+
+
+class Query(Model):
+
+    """ORM model for SQL query"""
+
+    __tablename__ = 'query'
+    id = Column(Integer, primary_key=True)
+    client_id = Column(String(11), unique=True, nullable=False)
+
+    database_id = Column(Integer, ForeignKey('dbs.id'), nullable=False)
+
+    # Store the tmp table into the DB only if the user asks for it.
+    tmp_table_name = Column(String(256))
+    user_id = Column(
+        Integer, ForeignKey('ab_user.id'), nullable=True)
+    status = Column(String(16), default=QueryStatus.PENDING)
+    tab_name = Column(String(256))
+    sql_editor_id = Column(String(256))
+    schema = Column(String(256))
+    sql = Column(Text)
+    # Query to retrieve the results,
+    # used only in case of select_as_cta_used is true.
+    select_sql = Column(Text)
+    executed_sql = Column(Text)
+    # Could be configured in the superset config.
+    limit = Column(Integer)
+    limit_used = Column(Boolean, default=False)
+    limit_reached = Column(Boolean, default=False)
+    select_as_cta = Column(Boolean)
+    select_as_cta_used = Column(Boolean, default=False)
+
+    progress = Column(Integer, default=0)  # 1..100
+    # # of rows in the result set or rows modified.
+    rows = Column(Integer)
+    error_message = Column(Text)
+    # key used to store the results in the results backend
+    results_key = Column(String(64), index=True)
+
+    # Using Numeric in place of DateTime for sub-second precision
+    # stored as seconds since epoch, allowing for milliseconds
+    start_time = Column(Numeric(precision=3))
+    start_running_time = Column(Numeric(precision=3))
+    end_time = Column(Numeric(precision=3))
+    changed_on = Column(
+        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=True)
+
+    database = relationship(
+        'Database',
+        foreign_keys=[database_id],
+        backref=backref('queries', cascade='all, delete-orphan')
+    )
+    user = relationship(
+        'User',
+        foreign_keys=[user_id])
+
+    __table_args__ = (
+        sqla.Index('ti_user_id_changed_on', user_id, changed_on),
+    )
+
+    @property
+    def limit_reached(self):
+        return self.rows == self.limit if self.limit_used else False
+
+    def to_dict(self):
+        return {
+            'changedOn': self.changed_on,
```

Comment:
Just moving the class from a module to another here as an effort to break down `models.py` that was humongous, we should get to the bottom of it in another PR though.